File: api-legacy/src/main/java/com/bitstachio/bazaarapi/common/exception/GlobalExceptionHandler.java
package com.bitstachio.bazaarapi.common.exception;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(ResourceNotFoundException.class)
    public ResponseEntity<ErrorResponse> handleResourceNotFound(ResourceNotFoundException ex) {
        var response =
                new ErrorResponse(
                        HttpStatus.NOT_FOUND.value(), ex.getMessage(), LocalDateTime.now());

        return new ResponseEntity<>(response, HttpStatus.NOT_FOUND);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponse> handleGlobalException() {
        var response =
                new ErrorResponse(
                        HttpStatus.INTERNAL_SERVER_ERROR.value(),
                        "An unexpected error occurred",
                        LocalDateTime.now());

        return new ResponseEntity<>(response, HttpStatus.INTERNAL_SERVER_ERROR);
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<Map<String, String>> handleValidationExceptions(
            MethodArgumentNotValidException ex) {
        var errors = new HashMap<String, String>();
        ex.getBindingResult()
                .getFieldErrors()
                .forEach(error -> errors.put(error.getField(), error.getDefaultMessage()));

        return new ResponseEntity<>(errors, HttpStatus.BAD_REQUEST);
    }
}

File: api/src/main/kotlin/dev/bitstachio/interdimensional_bazaar/common/exception/GlobalExceptionHandler.kt
package dev.bitstachio.interdimensional_bazaar.common.exception

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import java.time.LocalDateTime
import dev.bitstachio.interdimensional_bazaar.product.exception.DuplicateReviewException

@RestControllerAdvice
class GlobalExceptionHandler {

	@ExceptionHandler(ResourceNotFoundException::class)
	fun handleResourceNotFound(ex: ResourceNotFoundException): ResponseEntity<ErrorResponse> {
		val response =
			ErrorResponse(HttpStatus.NOT_FOUND.value(), ex.message ?: "", LocalDateTime.now())
		return ResponseEntity(response, HttpStatus.NOT_FOUND)
	}

	@ExceptionHandler(BadRequestException::class)
	fun handleBadRequest(ex: BadRequestException): ResponseEntity<ErrorResponse> {
		val response =
			ErrorResponse(HttpStatus.BAD_REQUEST.value(), ex.message ?: "", LocalDateTime.now())
		return ResponseEntity(response, HttpStatus.BAD_REQUEST)
	}

	@ExceptionHandler(DuplicateReviewException::class)
	fun handleDuplicateReview(ex: DuplicateReviewException): ResponseEntity<ErrorResponse> {
		val response = ErrorResponse(
			HttpStatus.CONFLICT.value(),
			ex.message ?: "",
			LocalDateTime.now(),
		)
		return ResponseEntity(response, HttpStatus.CONFLICT)
	}

	@ExceptionHandler(Exception::class)
	fun handleGlobalException(): ResponseEntity<ErrorResponse> {
		val response =
			ErrorResponse(
				HttpStatus.INTERNAL_SERVER_ERROR.value(),
				"An unexpected error occurred",
				LocalDateTime.now(),
			)
		return ResponseEntity(response, HttpStatus.INTERNAL_SERVER_ERROR)
	}

	@ExceptionHandler(MethodArgumentNotValidException::class)
	fun handleValidationExceptions(
		ex: MethodArgumentNotValidException,
	): ResponseEntity<Map<String, String>> {
		val errors = HashMap<String, String>()
		ex.bindingResult.fieldErrors.forEach { error ->
			errors[error.field] = error.defaultMessage ?: ""
		}
		return ResponseEntity(errors, HttpStatus.BAD_REQUEST)
	}
}
